File: backend/application/resume_service.py
import json
from uuid import UUID

from domain.resume import Resume, ResumeEducation, ResumeExperience, ResumeProject
from infrastructure.repositories.base import ResumeRepositoryInterface


class ResumeService:
    def __init__(self, repository: ResumeRepositoryInterface):
        self._repo = repository

    async def create_resume(self, user_id: UUID, data: dict) -> Resume:
        resume = Resume(
            user_id=user_id,
            title=data.get("title", "Untitled Resume"),
            personal_info=data.get("personal_info", {}),
            skills=data.get("skills", ""),
            template=data.get("template", "classic"),
            section_order=data.get("section_order", ["education", "skills", "experience", "projects"]),
            education=[
                ResumeEducation(
                    college=e.get("college", ""),
                    degree=e.get("degree", ""),
                    cgpa=e.get("cgpa", ""),
                    start_year=e.get("startYear", e.get("start_year", "")),
                    end_year=e.get("endYear", e.get("end_year", "")),
                )
                for e in data.get("education", [])
            ],
            experience=[
                ResumeExperience(
                    company=e.get("company", ""),
                    role=e.get("role", ""),
                    description=e.get("description", ""),
                )
                for e in data.get("experience", [])
            ],
            projects=[
                ResumeProject(
                    name=p.get("name", ""),
                    technologies=p.get("technologies", ""),
                    description=p.get("description", ""),
                )
                for p in data.get("projects", [])
            ],
        )
        return await self._repo.create(resume)

    async def get_resume(self, resume_id: UUID, user_id: UUID) -> Resume | None:
        return await self._repo.get(resume_id, user_id)

    async def list_resumes(self, user_id: UUID) -> list[dict]:
        return await self._repo.list_by_user(user_id)

    async def update_resume(self, resume_id: UUID, user_id: UUID, data: dict) -> Resume | None:
        existing = await self._repo.get(resume_id, user_id)
        if not existing:
            return None

        existing.title = data.get("title", existing.title)
        existing.personal_info = data.get("personal_info", existing.personal_info)
        existing.skills = data.get("skills", existing.skills)
        existing.template = data.get("template", existing.template)
        existing.section_order = data.get("section_order", existing.section_order)

        if "education" in data:
            existing.education = [
                ResumeEducation(
                    id=UUID(e.get("id")) if e.get("id") else None,
                    resume_id=resume_id,
                    sort_order=i,
                    college=e.get("college", ""),
                    degree=e.get("degree", ""),
                    cgpa=e.get("cgpa", ""),
                    start_year=e.get("startYear", e.get("start_year", "")),
                    end_year=e.get("endYear", e.get("end_year", "")),
                )
                for i, e in enumerate(data["education"])
            ]

        if "experience" in data:
            existing.experience = [
                ResumeExperience(
                    id=UUID(e.get("id")) if e.get("id") else None,
                    resume_id=resume_id,
                    sort_order=i,
                    company=e.get("company", ""),
                    role=e.get("role", ""),
                    description=e.get("description", ""),
                )
                for i, e in enumerate(data["experience"])
            ]

        if "projects" in data:
            existing.projects = [
                ResumeProject(
                    id=UUID(p.get("id")) if p.get("id") else None,
                    resume_id=resume_id,
                    sort_order=i,
                    name=p.get("name", ""),
                    technologies=p.get("technologies", ""),
                    description=p.get("description", ""),
                )
                for i, p in enumerate(data["projects"])
            ]

        return await self._repo.update(existing)

    async def delete_resume(self, resume_id: UUID, user_id: UUID) -> bool:
        return await self._repo.delete(resume_id, user_id)

    def reconstruct_resume_text(self, resume: Resume) -> str:
        """Convert structured resume to plain text for RAG ingestion."""
        return resume.to_text()

File: backend/src/services/resume-service.js
import { Resume, ResumeEducation, ResumeExperience, ResumeProject } from "../models/resume.js";

export class ResumeService {
  constructor(repository) {
    this.repository = repository;
  }

  async createResume(userId, data) {
    const resume = new Resume({
      userId,
      title: data.title || "Untitled Resume",
      personalInfo: data.personal_info || {},
      skills: data.skills || "",
      template: data.template || "classic",
      sectionOrder: data.section_order || [
        "education",
        "skills",
        "experience",
        "projects",
      ],
      education: (data.education || []).map(
        (e) =>
          new ResumeEducation({
            college: e.college || "",
            degree: e.degree || "",
            cgpa: e.cgpa || "",
            startYear: e.startYear || e.start_year || "",
            endYear: e.endYear || e.end_year || "",
          }),
      ),
      experience: (data.experience || []).map(
        (e) =>
          new ResumeExperience({
            company: e.company || "",
            role: e.role || "",
            description: e.description || "",
          }),
      ),
      projects: (data.projects || []).map(
        (p) =>
          new ResumeProject({
            name: p.name || "",
            technologies: p.technologies || "",
            description: p.description || "",
          }),
      ),
    });
    return this.repository.create(resume);
  }

  async getResume(resumeId, userId) {
    return this.repository.get(resumeId, userId);
  }

  async listResumes(userId) {
    return this.repository.listByUser(userId);
  }

  async updateResume(resumeId, userId, data) {
    const existing = await this.repository.get(resumeId, userId);
    if (!existing) return null;

    existing.title = data.title ?? existing.title;
    existing.personalInfo = data.personal_info ?? existing.personalInfo;
    existing.skills = data.skills ?? existing.skills;
    existing.template = data.template ?? existing.template;
    existing.sectionOrder = data.section_order ?? existing.sectionOrder;

    if ("education" in data) {
      existing.education = (data.education || []).map(
        (e, i) =>
          new ResumeEducation({
            id: e.id || undefined,
            resumeId,
            sortOrder: i,
            college: e.college || "",
            degree: e.degree || "",
            cgpa: e.cgpa || "",
            startYear: e.startYear || "",
            endYear: e.endYear || "",
          }),
      );
    }

    if ("experience" in data) {
      existing.experience = (data.experience || []).map(
        (e, i) =>
          new ResumeExperience({
            id: e.id || undefined,
            resumeId,
            sortOrder: i,
            company: e.company || "",
            role: e.role || "",
            description: e.description || "",
          }),
      );
    }

    if ("projects" in data) {
      existing.projects = (data.projects || []).map(
        (p, i) =>
          new ResumeProject({
            id: p.id || undefined,
            resumeId,
            sortOrder: i,
            name: p.name || "",
            technologies: p.technologies || "",
            description: p.description || "",
          }),
      );
    }

    return this.repository.update(existing);
  }

  async deleteResume(resumeId, userId) {
    return this.repository.delete(resumeId, userId);
  }

  reconstructResumeText(resume) {
    return resume.toText();
  }
}
